### PR TITLE
unit: cancel conflict handler

### PR DESCRIPTION
**Pull request by jayahariv**
* check with/without update to the current document if we return false from the conflict handler, how the save behaves.

ref: #2430

### CODE REVIEW

File: Objective-C/Tests/DatabaseTest.m [hunk 1]
```diff
@@ -582,6 +582,57 @@ - (void) testSaveDocWithDeletedConflictUsingConcurrencyControl: (int)concurrency
...
+    CBLMutableDocument* doc1b = [[self.db documentWithID: docID] toMutable];
+    
+    [doc1a setString: @"Scotty" forKey: @"nickName"];
+    [self saveDocument: doc1a];
+    AssertEqual([self.db documentWithID: docID].generation, 2u);
+    
+    NSError* error;
+    [doc1b setString: @"Scott" forKey: @"nickName"];
+    [self.db saveDocument: doc1b
+          conflictHandler:^BOOL(CBLMutableDocument * cur, CBLDocument * old) {
```
Comment (snej) [inline Objective-C/Tests/DatabaseTest.m]: The names `cur` and `old` confused me -- `old` isn't old, it's the newer revision in the database. So I had them backwards. Maybe `savingDoc` and `dbDoc` or something like that?

File: Objective-C/Tests/DatabaseTest.m [hunk 2]
```diff
@@ -582,6 +582,57 @@ - (void) testSaveDocWithDeletedConflictUsingConcurrencyControl: (int)concurrency
...
+              AssertEqualObjects(doc1b.toDictionary, cur.toDictionary);
+              AssertEqualObjects(doc1a.toDictionary, old.toDictionary);
+              return NO;
+          } error: &error];
+    AssertNil(error);
```
Comment (snej) [inline Objective-C/Tests/DatabaseTest.m]: Isn't `saveDocument` supposed to return a Conflict error if the handler returns NO?

Also, testing `error` is the wrong way to check for success/failure. In Cocoa APIs, the return value indicates success/failure. The NSError has a value only on failure; on success _its value is undefined_. (Yeah, you initialized it to nil, but it's possible that the method you called used it itself for other calls that failed, but it recovered from those and returned YES, but left the error non-nil.)

Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m]: will assert the `db saveDocument` method. But should we also make sure, the provided error is empty when success

Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m]: yea, this was wrong, checking error as empty. which is corrected.

File: Objective-C/Tests/DatabaseTest.m (hunk 1 above)
Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m]: As per the design spec this variables are `document` and `oldDocument`. We can discuss about this when Pasin is back.

Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m]: corrected to use the design-spec, with `document` and `oldDocument`

**Pull request merged**